hoodie-spark/src/main/scala/com/uber/hoodie/DefaultSource.scala

```diff
@@ -177,93 +192,107 @@ class DefaultSource extends RelationProvider
-      properties.put(HoodieTableConfig.HOODIE_TABLE_TYPE_PROP_NAME, storageType);
-      properties.put(HoodieTableConfig.HOODIE_ARCHIVELOG_FOLDER_PROP_NAME, "archived");
-      HoodieTableMetaClient.initializePathAsHoodieDataset(sparkContext.hadoopConfiguration, path.get, properties);
-    }
-
-    // Create a HoodieWriteClient & issue the write.
-    val jsc = new JavaSparkContext(sparkContext);
-    val client = DataSourceUtils.createHoodieClient(jsc,
-      schema.toString,
-      path.get,
-      tblName.get,
-      mapAsJavaMap(parameters)
-    )
-    val commitTime = client.startCommit();
-
-    val writeStatuses = DataSourceUtils.doWriteOperation(client, hoodieRecords, commitTime, operation)
-    // Check for errors and commit the write.
-    val errorCount = writeStatuses.rdd.filter(ws => ws.hasErrors).count()
-    if (errorCount == 0) {
-      log.info("No errors. Proceeding to commit the write.");
-      val metaMap = parameters.filter(kv =>
-        kv._1.startsWith(parameters(COMMIT_METADATA_KEYPREFIX_OPT_KEY)))
-      val success = if (metaMap.isEmpty) {
-        client.commit(commitTime, writeStatuses)
+    val hoodieRecords =
+      if (parameters(INSERT_DROP_DUPS_OPT_KEY).toBoolean) {
+        DataSourceUtils.dropDuplicates(
+          jsc,
+          hoodieAllIncomingRecords,
+          mapAsJavaMap(parameters))
       } else {
-        client.commit(commitTime, writeStatuses,
-          Optional.of(new util.HashMap[String, String](mapAsJavaMap(metaMap))))
+        hoodieAllIncomingRecords
       }
 
-      if (success) {
-        log.info("Commit " + commitTime + " successful!")
+    if (!hoodieRecords.isEmpty()) {
+      val basePath = new Path(parameters.get("path").get)
+      val fs = basePath.getFileSystem(sparkContext.hadoopConfiguration)
+      var exists = fs.exists(basePath)
+
+      // Handle various save modes
+      if (mode == SaveMode.ErrorIfExists && exists) {
+        throw new HoodieException(s"basePath ${basePath} already exists.")
+      }
+      if (mode == SaveMode.Ignore && exists) {
+        log.warn(s" basePath ${basePath} already exists. Ignoring & not performing actual writes.")
+        return createRelation(sqlContext, parameters, df.schema)
       }
-      else {
-        log.info("Commit " + commitTime + " failed!")
+      if (mode == SaveMode.Overwrite && exists) {
+        log.warn(s" basePath ${basePath} already exists. Deleting existing data & overwriting with new data.")
+        fs.delete(basePath, true)
+        exists = false
       }
 
-      val hiveSyncEnabled = parameters.get(HIVE_SYNC_ENABLED_OPT_KEY).map(r => r.toBoolean).getOrElse(false)
-      if (hiveSyncEnabled) {
-        log.info("Syncing to Hive Metastore (URL: " + parameters(HIVE_URL_OPT_KEY) + ")")
-        val fs = FSUtils.getFs(basePath.toString, jsc.hadoopConfiguration)
-        syncHive(basePath, fs, parameters)
+      // Create the dataset if not present (APPEND mode)
+      if (!exists) {
```

Comment:
headsup : you may run into conflicts when rebasing, since I pulled this block into a helper method